docs/source/generics.rst

```diff
@@ -0,0 +1,986 @@
+   class Another(Generic[T]): ...
+
+   class First(One[T], Another[S]): ...
+   class Second(One[T], Another[S], Generic[S, U, T]): ...
+
+   x: First[int, str]        # Here T is bound to int, S is bound to str
+   y: Second[int, str, Any]  # Here T is Any, S is int, and U is str
+
+.. _generic-functions:
+
+Generic functions
+*****************
+
+Type variables can be used to define generic functions. These are functions
+where the types of the arguments or return value have some relationship:
+
+.. code-block:: python
+
+   from typing import TypeVar, Sequence
+
+   T = TypeVar('T')
+
+   # A generic function!
+   def first(seq: Sequence[T]) -> T:
+       return seq[0]
+
+As with generic classes, the type variable can be replaced with any
+type. That means ``first`` can be used with any sequence type, and the
+return type is derived from the sequence item type. For example:
+
+.. code-block:: python
+
+   reveal_type(first([1, 2, 3]))   # Revealed type is "builtins.int"
+   reveal_type(first(['a', 'b']))  # Revealed type is "builtins.str"
+
+Since type variables are about describing the relationship between
+two or more types, it's usually not useful to have a type variable
+only appear once in a function signature.
+
+Note that for convenience, a single type variable symbol (such as ``T`` above)
+can be used in multiple generic functions or classes, even though the logical
+scope is different in each generic function or class. In the following example
+we reuse the same type variable symbol in two generic functions; these two
+functions do not share any typing relationship to each other:
+
+.. code-block:: python
+
+   from typing import TypeVar, Sequence
+
+   T = TypeVar('T')
+
+   def first(seq: Sequence[T]) -> T:
+       return seq[0]
+
+   def last(seq: Sequence[T]) -> T:
+       return seq[-1]
+
+Variables should not have a type variable in its type unless the type variable
+is bound by a containing generic class, generic function or generic alias.
+
+.. _generic-methods-and-generic-self:
+
+Generic methods and generic self
+********************************
+
+You can also define generic methods — just use a type variable in the
+method signature that is different from the type variable(s) bound in
+the class definition.
+
+.. code-block:: python
+
+    # T is the type variable bound by this class
+    class PairedBox(Generic[T]):
+        def __init__(self, content: T) -> None:
+            self.content = content
+
+        # S is a type variable bound only in this method
+        def first(self, x: list[S]) -> S:
+            return x[0]
+
+        def pair_with_first(self, x: list[S]) -> tuple[S, T]:
+            return (x[0], self.content)
+
+    box = PairedBox("asdf")
+    reveal_type(box.first([1, 2, 3]))  # Revealed type is "builtins.int"
+    reveal_type(box.pair_with_first([1, 2, 3]))  # Revealed type is "tuple[builtins.int, builtins.str]"
+
+In particular, the ``self`` argument may also be generic, allowing a
```

Comment:
This example isn't great as it should be using `Self`. Not sure if there's a better example to give, though.